spark/src/main/scala/org/apache/spark/sql/delta/expressions/HilbertLongIndex.scala

```diff
@@ -0,0 +1,401 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.expressions
+
+import java.util
+
+import scala.collection.mutable
+
+import org.apache.spark.sql.delta.expressions.HilbertUtils._
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{ExpectsInputTypes, Expression}
+import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
+import org.apache.spark.sql.types.{AbstractDataType, DataType, DataTypes}
+
+/**
+ * Represents a hilbert index built from the provided columns.
+ * The columns are expected to all be Ints and to have at most numBits.
+ * The points along the hilbert curve are represented by Longs.
+ */
+private[sql] case class HilbertLongIndex(numBits: Int, children: Seq[Expression])
```

Comment:
Is there a maximum practical value of numBits and/or a maximum practical number of children given we're targeting a 64-bit space of output values?